PreExecution non-repudiation implementation

tdimitrov opened the pull request:
This PR implements PreExecution result verification by non-primary replicas (aka PreExecution non-repudiation).

What it does:
1. Adds flag to the PreProcessor indicating if the feature should be enabled or not (feature flag).
2. Adds new parameter to ClientRequestMessage, which allows allocating extra buffer space by MessageBase class. This buffer can be used by messages inheriting from ClientRequestMessage to store additional fields (see next point).
3. Adds new PreExecution message - PreProcessedResult message. It is carries the result from the PreExecution and the signatures of all f+1 replicas which has preexecuted the concrete request. This message is implemented by inheriting from ClientRequestMessage. The feature described in point 2 is used to allocate additional buffer space for the signatures.
4. Result verification logic handling.

How the result verification logic changes the PreExecution flow?
================================================
The flow remains unchanged up to receiving a PreProcessReply message on the primary. Now instead of adding a ClientRequestMsg to its internal queue the primary uses PreProcessedResult for this purpose. A clear benefit from this is responsibility separation - ClientRequest messages are used for client requests and PreProcessedResult - for pre-executed requests. All hash signatures from the non-primary replicas are added in the PreProcessedResult message. All its other fields are 1:1 compared to ClientRequest.

On finalise preprocessing phase (finalizePreprocessing method) the Primary bundles some ClientRequest and PreProcessedResult messages in a PrePrepare message. There is another benefit from the inheritance here - PrePrepare handling logic doesn't care about the new message. From its point of view it is just another ClientRequest message.

When a non-primary receives a PrePrepare message it unbundles all ClientRequest and PreProcessedResult messages from it. They are discriminated by the message type in the message header. If the non-primary find a PreProcessedResult message it verifies the validity of all signatures in it and additionally if f+1 signatures are assigned. If both requirements are satisfied the message is processed as usual.

However if a non-primary finds a PreProcessedResult message with invalid signature or less than f+1 signatures it assumes that the primary is malicious and triggers a view change. New reason is added to ReplicaAsksToLeaveViewMsg -  PrimarySentBadPreProcessedResult.

tdimitrov commented:
@yuliasherman @arc-vmware this is still work in progress. I suggest to delay you review with a few days.

yuliasherman commented on bftengine/src/preprocessor/RequestProcessingState.cpp:
```diff
@@ -103,7 +103,10 @@ void RequestProcessingState::handlePreProcessReplyMsg(const PreProcessReplyMsgSh
+    // Counts equal hashes and saves the signatures with the replica ID. They will be used as a proof that the primary
+    // is sending correct preexecution result to the rest of the replicas.
+    preProcessingResultHashes_[newHashArray].emplace_back(preProcessReplyMsg->getResultHashSignature(),
+                                                          preProcessReplyMsg->senderId());
```
Didn't you lose "++"?

tdimitrov replied:
Sorry, but I don't follow. Which value should be incremented?

yuliasherman replied:
Forget about it, it's OK.

yuliasherman commented on bftengine/src/preprocessor/RequestProcessingState.cpp:
```diff
@@ -199,4 +202,11 @@ unique_ptr<MessageBase> RequestProcessingState::buildClientRequestMsg(bool empty
   return clientPreProcessReqMsg_->convertToClientRequestMsg(emptyReq);
 }
 
+// Returns const reference to internal state. NOT THREAD SAFE!
```
Actually, none of RequestProcessingState functions is thread-safe. Maybe worth adding a general comment that they are guarded by RequestState::mutex from PreProcessor.

yuliasherman commented on bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:
```diff
@@ -0,0 +1,106 @@
+// Concord
+//
+// Copyright (c) 2018-2021 VMware, Inc. All Rights Reserved.
```
Nit: completely new file: 2021

yuliasherman commented on bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:
```diff
@@ -0,0 +1,106 @@
+                                             const concordUtils::SpanContext& spanContext,
+                                             const char* messageSignature,
+                                             uint32_t messageSignatureLen,
+                                             const char* resultSignaturesBuf,
```
Consider passing a string instead.

yuliasherman commented on bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:
```diff
@@ -0,0 +1,106 @@
+    signature_size = concordUtils::fromBigEndianBuffer<uint32_t>(buf + pos);
+    pos += sizeof(uint32_t);
+
+    if (signature_size > len - pos) {
```
This should be verified in validate() function, which looks not implemented.

tdimitrov replied:
This check is "deferred" due to the VC trigger we talked about. This is the only reason not to put it in validate().

Is this acceptable after the validation code is removed from ReplicaImp?

yuliasherman replied:
Yes, thanks.

yuliasherman commented on bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:
```diff
@@ -0,0 +1,106 @@
+                                                                                               size_t len) {
+  size_t pos = 0;
+  std::list<PreProcessResultSignature> ret;
+  while (1) {
```
We should know ahead how long is the signatures list. The signature size is also constant.

tdimitrov replied:
I agree that we should know that there will be f signatures and its size is constant, but I have a reason to do it in such a way.

I put the signature lengths so that the message encoding/decoding is decoupled from the SIgManager. I think it's better to pass value sizes in the message unless they are a hard constant. In this case I don't consider them hard constant, because if we change the hashing algorithm this size will change. Replicas will easily figure this out, but the message will be unparsable outside Concord.

My motivation about the list size is similar - we can do for loop from 0 to F, but I wanted to implement more independant message format. I could have put the list size in the message too, but it seemed redundant for me.

yuliasherman commented on bftengine/src/preprocessor/messages/PreProcessedResultMsg.hpp:
```diff
@@ -0,0 +1,73 @@
+// Concord
+//
+// Copyright (c) 2018-2021 VMware, Inc. All Rights Reserved.
```
->2021

yuliasherman reviewed (commented):
Great job!

tdimitrov commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -668,6 +681,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
 
+bool ReplicaImp::validatePreProcessedResultSignatures(preprocessor::PreProcessedResultMsg &msg) {
+  auto [buf, buf_len] = msg.getResultSignaturesBuf();
+  auto sigs = preprocessor::PreProcessResultSignature::deserializeResultSignatureList(buf, buf_len);
```
RVO kicks in here so no copy is made. We can't capture the return by reference, because it's a local variable inside deserializeResultSignatureList().

yuliasherman replied:
OK, I've missed that we return a local variable.

arc-vmware commented on bftengine/include/bftengine/ClientMsgs.hpp:
```diff
@@ -40,6 +40,7 @@ struct ClientRequestMsgHeader {
   uint64_t timeoutMilli;
   uint32_t cidLength = 0;
   uint16_t reqSignatureLength = 0;
+  uint32_t extraDataLength = 0;
```
Will this make Client request message backward incompatible. We should check if this addition effects serialisation and deserialisation of this message.

tdimitrov replied:
Actually this change doesn't affect the ClientRequestMsg. It's a mechanism to allocate extra buffer space if you are inheriting from the message (as it's done for PreProcess(ed)Result message).

The ClientRequestMsg message itself is not affected by this change because the extraDataLength is always set to 0 for it.

I also had concerns for serialisation/deserialisation so I added additional test cases in the message unit test. Check my changes in ClientRequestMsg_test.cpp/

arc-vmware replied:
Checked the unit test, I think this should work.

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+  }
+
+  for (const auto &s : sigs) {
+    auto hash = concord::util::SHA3_256().digest(msg.requestBuf(), msg.requestLength());
```
We should create the hashes in parallel. Its done in PrePrepareMsg creation.

tdimitrov replied:
Good point, I'll do it.

tdimitrov replied:
Actually the hash is calculated only once and should be outside the for loop. Good catch!

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+    auto hash = concord::util::SHA3_256().digest(msg.requestBuf(), msg.requestLength());
+    bool verificationResult = false;
+    if (getReplicaConfig().replicaId == s.sender_replica) {
+      std::vector<char> mySignature(sigManager_->getMySigLength(), 0);
```
its better to pass '\0' instead of 0 as the second parameter is a char.

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+    bool verificationResult = false;
+    if (getReplicaConfig().replicaId == s.sender_replica) {
+      std::vector<char> mySignature(sigManager_->getMySigLength(), 0);
+      sigManager_->sign((const char *)hash.data(), hash.size(), mySignature.data(), mySignature.size());
```
we should static_cast instead of the c-style cast.

tdimitrov replied:
We can't static_cast from unsigned to signed, so I'll replace it with reinterpret_cast.

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+    if (getReplicaConfig().replicaId == s.sender_replica) {
+      std::vector<char> mySignature(sigManager_->getMySigLength(), 0);
+      sigManager_->sign((const char *)hash.data(), hash.size(), mySignature.data(), mySignature.size());
+      verificationResult = mySignature == s.signature;
```
if replica himself gets his own message, then will there be any chance of changes? If no then we can just avoid checking the signature for same replica.

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+bool ReplicaImp::validatePrePrepareMsg(const PrePrepareMsg *msg, const bftEngine::impl::ReplicasInfo &replicasInfo) {
+  RequestsIterator reqIter(msg);
+  char *requestBody = nullptr;
+  while (reqIter.getAndGoToNext(requestBody)) {
```
The validate() method of PrePrepareMsg is doing the same. We can call that function here.

arc-vmware commented on bftengine/src/bftengine/ReplicaImp.hpp:
```diff
@@ -480,6 +483,9 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
   bool isSeqNumToStopAt(SeqNum seq_num);
 
+  bool validatePrePrepareMsg(const PrePrepareMsg* msg, const bftEngine::impl::ReplicasInfo& replicasInfo);
+  bool validatePreProcessedResultSignatures(preprocessor::PreProcessedResultMsg& msg);
```
As this is validate [read only] method, should msg be const& ?

tdimitrov replied:
Good point. I tried this while writing the function. There is a call to msg.getResultSignaturesBuf() in it:
`const auto [buf, buf_len] = msg.getResultSignaturesBuf();`
getResultSignaturesBuf returns pointer so it's not correct to be const. This was the only reason to  get non-const reference in validatePreProcessedResultSignatures().

arc-vmware commented on bftengine/src/bftengine/messages/ClientRequestMsg.hpp:
```diff
@@ -75,14 +76,20 @@ class ClientRequestMsg : public MessageBase {
   void validateImp(const ReplicasInfo& repInfo) const;
 
+  // Returns a pair of pointer and size to the extra buffer which was allocated during initialisation
+  std::pair<char*, uint32_t> getExtraBufPtr() {
```
nit: If we need only length then also this will result into buffer pointer calculation.

tdimitrov replied:
True, but we usually need both. Even if the pointer is not used I think it's a small price to pay.

arc-vmware commented on bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp:
```diff
@@ -87,6 +87,15 @@ void PreProcessReplyMsg::validate(const ReplicasInfo& repInfo) const {
   }
 }  // namespace preprocessor
 
+std::vector<char> PreProcessReplyMsg::getResultHashSignature() const {
```
nit: [very minor comment] void PreProcessReplyMsg::getResultHashSignature(std::vector<char> & result_hash) -- This will avoid result vector and copy ctors.

tdimitrov replied:
Actually no copy happens here too. RVO is applied for return-by-value vector. 

This function is called in RequestProcessingState::handlePreProcessReplyMsg(), as a parameter for emplace_back() which accepts universal reference. This is one of the cases where copy elision is applied so we are fine I think.

arc-vmware replied:
yeah, i missed the RVO. this is absolutely fine. Thanks @tdimitrov

arc-vmware commented on bftengine/src/preprocessor/PreProcessor.cpp:
```diff
@@ -800,7 +811,7 @@ void PreProcessor::cancelPreProcessing(NodeIdType clientId, uint16_t reqOffsetIn
 
 void PreProcessor::finalizePreProcessing(NodeIdType clientId, uint16_t reqOffsetInBatch) {
-  std::unique_ptr<ClientRequestMsg> clientRequestMsg;
+  std::unique_ptr<impl::MessageBase> preProcessedMsg;
```
nit: Nothing to change, just as assertion : if we keep this ClientRequestMsg then also it should be fine, since preprocess result msg is a type of ClientRequestMsg.

tdimitrov replied:
Initially we will keep both logics and switch between them via configuration. But later we have to remove the old approach (with ClientRequestMsg reuse), because it is not optimal.

arc-vmware reviewed (commented):
Great work, enjoyed and learned quite a lot. I have few minor comments, feel free to defer if they are not relevant.

tdimitrov commented on bftengine/src/preprocessor/PreProcessor.cpp:
```diff
@@ -812,21 +823,42 @@ void PreProcessor::finalizePreProcessing(NodeIdType clientId, uint16_t reqOffset
+                                                             span_context,
+                                                             reqProcessingStatePtr->getReqSignature(),
+                                                             reqProcessingStatePtr->getReqSignatureLength(),
+                                                             sigsBuf.data(),
```
I agree, will do it for sigsBuf too

tdimitrov commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -42,10 +42,12 @@
 #include "messages/ReplicaAsksToLeaveViewMsg.hpp"
 #include "messages/ReplicaRestartReadyMsg.hpp"
 #include "messages/ReplicasRestartReadyProofMsg.hpp"
+#include "messages/PreProcessedResultMsg.hpp"
```
It was painful, but I did it ;)

tdimitrov commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -668,6 +681,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+    return false;
+  }
+
+  for (const auto &s : sigs) {
```
My idea about PreProcess(ed)ResultMsg is to be just a carrier for the signatures without actually validating them.

Initially I wanted to have a validate() method which does the signature checks. However this validate is called very early in the message processing code and if it fails the message is discarded. This doesn't work for us because we want to initiate a view change if signature verification fails. 

After considering pros and cons I decided that it's better to handle this in ReplicaImp because (1) it makes more sense to have a view change triggering logic there and (2) I didn't want to couple PreProcess(ed)ResultMsg with SigManager too much. I think this will bite us in the future when we decide to rework the message encoding/decoding code.

Does this make sense?

yuliasherman replied:
We all the time trying to offload the ReplicaImp code from a non-essential functionally, but the functions you added break this approach. I think it's less evil to have PreProcessResultMsg be aware of SigManager than adding all this functionality to ReplicaImp. In case a VC is required, the corresponding PreProcessResultMsg function needs just to return a flag.

tdimitrov replied:
Yes, message validation can be separated from ViewChange triggering. I've fixed this.

tdimitrov commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -668,6 +681,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
+  return true;
+}
+
+bool ReplicaImp::validatePrePrepareMsg(const PrePrepareMsg *msg, const bftEngine::impl::ReplicasInfo &replicasInfo) {
```
Please check my previous comment: https://github.com/vmware/concord-bft/pull/1639#discussion_r665335187

tdimitrov replied:
This is also fixed.

tdimitrov commented:
Thanks for the feedback @yuliasherman @arc-vmware . I have addressed your comments. There are a bunch of commits which I will squash before merging.
Could you please approve the PR if you are satisfied with the changes?

yuliasherman reviewed (approved):
LGTM, 10x!

arc-vmware reviewed (approved):
Awesome, great feature getting in for concord.
